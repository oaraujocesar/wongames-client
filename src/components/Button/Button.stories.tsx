import { ComponentStory, ComponentMeta } from '@storybook/react'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'
import Button from './Button'

export default {
  title: 'Button',
  component: Button,
  args: {
    children: {
      type: 'string'
    }
  }
} as ComponentMeta<typeof Button>

export const Default: ComponentStory<typeof Button> = (args) => (
  <Button {...args} />
)

Default.args = {
  children: 'Buy now'
}

export const Icon: ComponentStory<typeof Button> = (args) => (
  <Button {...args} />
)

Icon.args = {
  size: 'small',
  children: 'Buy now',
  icon: <AddShoppingCart />
}
