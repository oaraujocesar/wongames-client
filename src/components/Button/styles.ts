import styled, { css, DefaultTheme } from 'styled-components'

import { ButtonProps } from './Button'

type WrapperProps = { hasIcon: boolean } & Pick<
  ButtonProps,
  'size' | 'fullWidth'
>

const wrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    height: 3rem;
    font-size: ${theme.font.sizes.xsmall};
  `,
  medium: (theme: DefaultTheme) => css`
    height: 4rem;
    font-size: ${theme.font.sizes.small};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,
  large: (theme: DefaultTheme) => css`
    height: 5rem;
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};
  `,
  fullWidth: () => css`
    width: 100%;

    span {
      width: inherit;
    }
  `,
  withIcon: (theme: DefaultTheme) => css`
    svg {
      width: 1.5rem;

      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size, fullWidth, hasIcon }) => css`
    align-items: center;
    display: inline-flex;
    justify-content: center;
    padding: ${theme.spacings.xxsmall};

    border: none;
    cursor: pointer;
    text-decoration: none;
    color: ${theme.colors.white};
    font-family: ${theme.font.family};
    border-radius: ${theme.border.radius};
    background: linear-gradient(180deg, #ff5f5f 0%, #f062c0 100%);

    &:hover {
      background: linear-gradient(180deg, #e35565 0%, #d958a6 100%);
    }

    > span {
      margin: 0 auto;
    }

    ${!!size && wrapperModifiers[size](theme)};

    ${!!fullWidth && wrapperModifiers.fullWidth()};

    ${!!hasIcon && wrapperModifiers.withIcon(theme)};
  `}
`
