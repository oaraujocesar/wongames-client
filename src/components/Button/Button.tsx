import { Wrapper } from './styles'

export type ButtonProps = {
  children?: React.ReactNode
  size?: 'small' | 'medium' | 'large'
  fullWidth?: boolean
  icon?: JSX.Element
  onClick?: () => (event: React.MouseEvent<HTMLButtonElement>) => void
}

const Button = ({
  children,
  size = 'medium',
  fullWidth = false,
  icon,
  ...props
}: ButtonProps) => (
  <Wrapper size={size} fullWidth={fullWidth} hasIcon={!!icon} {...props}>
    {icon}
    {!!children && <span>{children}</span>}
  </Wrapper>
)

export default Button
