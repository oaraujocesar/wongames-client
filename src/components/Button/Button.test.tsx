import { screen } from '@testing-library/react'
import Button from './Button'

import { renderWithTheme } from 'utils/tests/helpers'

import { AddShoppingCart } from '@styled-icons/material-outlined'

describe('<Button />', () => {
  it('should render the button with the passed label', () => {
    renderWithTheme(<Button>Click me!</Button>)

    expect(
      screen.getByRole('button', { name: /Click me!/i })
    ).toBeInTheDocument()
  })

  it('should should render the biggest button', () => {
    renderWithTheme(<Button size="large">Click me!</Button>)

    expect(screen.getByRole('button', { name: /Click me!/i })).toHaveStyle({
      height: '5rem',
      padding: '0.8rem 4.8rem',
      'font-size': '1.6rem'
    })
  })

  it('should should render the normal button as default', () => {
    const { container } = renderWithTheme(<Button>Click me!</Button>)

    expect(screen.getByRole('button', { name: /Click me!/i })).toHaveStyle({
      height: '4rem',
      padding: '0.8rem 3.2rem',
      'font-size': '1.4rem'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should should render the smallest button', () => {
    renderWithTheme(<Button size="small">Click me!</Button>)

    expect(screen.getByRole('button', { name: /Click me!/i })).toHaveStyle({
      height: '3rem',
      padding: '0.8rem',
      'font-size': '1.2rem'
    })
  })

  it('should should render the full width button', () => {
    renderWithTheme(<Button fullWidth>Click me!</Button>)

    expect(screen.getByRole('button', { name: /Click me!/i })).toHaveStyle({
      width: '100%'
    })
  })

  it('should should render a button with an icon', () => {
    renderWithTheme(
      <Button icon={<AddShoppingCart data-testid="icon" />}>Click me!</Button>
    )

    expect(screen.getByText(/click me!/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('should render Button as a link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Click me!
      </Button>
    )

    expect(screen.getByRole('link', { name: /click me!/i })).toHaveAttribute(
      'href',
      '/link'
    )
  })
})
