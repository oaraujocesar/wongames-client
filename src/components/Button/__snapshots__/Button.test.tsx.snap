// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should should render the normal button as default 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.8rem;
  border: none;
  cursor: pointer;
  color: #FAFAFA;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  border-radius: 0.4rem;
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

<button
  class="c0"
>
  <span>
    Click me!
  </span>
</button>
`;
